sql/logictest: TestParallel failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/22119c85d2e7aca50f21f61408ae55207679569d

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestParallel PKG=github.com/cockroachdb/cockroach/pkg/sql/logictest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1140728&tab=buildLog

```
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:496 +0x20c
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendPartialBatch()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1139 +0x422
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).divideAndSendBatchToRanges()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:798 +0x19ae
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:710 +0x6ea
  github.com/cockroachdb/cockroach/pkg/internal/client.lookupRangeFwdScan()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/range_lookup.go:318 +0x5be
  github.com/cockroachdb/cockroach/pkg/internal/client.RangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/range_lookup.go:201 +0x7b4
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).RangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:343 +0xba
  github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).performRangeLookup()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:412 +0x400
  github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).lookupRangeDescriptorInternal.func3()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:285 +0x166
  github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight.(*Group).doCall()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight/singleflight.go:118 +0x4c

Goroutine 148 (running) created at:
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:314 +0x14fb
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.serveConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:167 +0x275
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:509 +0xc0d
  github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func20.1()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:1710 +0x1b7
  github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:139 +0xdf
==================
--- done: testdata/parallel_test/subquery_retry/txn with config : 1 tests, 0 failures
I190215 08:44:48.972145 396 sql/logictest/logic.go:907  --- done: testdata/parallel_test/subquery_retry/txn with config : 1 tests, 0 failures
--- done: testdata/parallel_test/subquery_retry/txn with config : 1 tests, 0 failures
I190215 08:44:49.059996 487 sql/logictest/logic.go:907  --- done: testdata/parallel_test/subquery_retry/txn with config : 1 tests, 0 failures
--- done: testdata/parallel_test/subquery_retry/txn with config : 1 tests, 0 failures
I190215 08:44:49.205905 438 sql/logictest/logic.go:907  --- done: testdata/parallel_test/subquery_retry/txn with config : 1 tests, 0 failures
I190215 08:44:49.206114 39 sql/logictest/parallel_test.go:159  testdata/parallel_test/subquery_retry/test.yaml: run list 2: 0:final
--- done: testdata/parallel_test/subquery_retry/final with config : 2 tests, 0 failures
I190215 08:44:49.225712 508 sql/logictest/logic.go:907  --- done: testdata/parallel_test/subquery_retry/final with config : 2 tests, 0 failures
I190215 08:44:49.226140 509 util/stop/stopper.go:547  quiescing; tasks left:
1      [async] transport racer
1      [async] intent_resolver_batcher
1      [async] closedts-rangefeed-subscriber
I190215 08:44:49.226512 509 util/stop/stopper.go:547  quiescing; tasks left:
1      [async] transport racer
1      [async] closedts-rangefeed-subscriber
I190215 08:44:49.228329 509 util/stop/stopper.go:547  quiescing; tasks left:
1      [async] transport racer
I190215 08:44:49.293822 166 kv/transport_race.go:113  transport race promotion: ran 42 iterations on up to 1147 requests
    --- FAIL: TestParallel/subquery_retry (3.43s)
        testing.go:771: race detected during execution of test

```

--- a/pkg/kv/txn_coord_sender.go
+++ b/pkg/kv/txn_coord_sender.go
@@ -480,7 +480,8 @@ func (tcf *TxnCoordSenderFactory) TransactionalSender(
 			// "refreshed away" without the need for a txn restart. Because the
 			// span refresher can re-issue batches, it needs to be careful about
 			// what parts of the batch it mutates. Any mutation needs to be
-			// idempotent.
+			// idempotent and should avoid writing to memory when not changing
+			// it to avoid looking like a data race.
 			&tcs.interceptorAlloc.txnCommitter,
 			// The metrics recorder sits at the bottom of the stack so that it
 			// can observe all transformations performed by other interceptors.

--- a/pkg/kv/txn_coord_sender.go
+++ b/pkg/kv/txn_coord_sender.go
@@ -488,19 +488,21 @@ func (tcf *TxnCoordSenderFactory) TransactionalSender(
 		}
 		tcs.interceptorStack = tcs.interceptorAlloc.arr[:]
 	case client.LeafTxn:
-		// LeafTxns never perform writes so the sequence number allocator should
-		// never increment its sequence number counter over its lifetime, but it
-		// still plays the important role of assigning each read request the
-		// latest sequence number.
-		tcs.interceptorAlloc.arr[0] = &tcs.interceptorAlloc.txnSeqNumAllocator
-		// The pipeliner is needed on leaves to ensure that in-flight writes are
-		// chained onto by reads that should see them.
-		tcs.interceptorAlloc.arr[1] = &tcs.interceptorAlloc.txnPipeliner
-		// The span refresher was configured above to not actually perform
-		// refreshes for leaves. It is still needed for accumulating the spans
-		// to be reported to the Root. But the gateway doesn't do much with
-		// them; see #24798.
-		tcs.interceptorAlloc.arr[2] = &tcs.interceptorAlloc.txnSpanRefresher
+		tcs.interceptorAlloc.arr = [cap(tcs.interceptorAlloc.arr)]txnInterceptor{
+			// LeafTxns never perform writes so the sequence number allocator
+			// should never increment its sequence number counter over its
+			// lifetime, but it still plays the important role of assigning each
+			// read request the latest sequence number.
+			&tcs.interceptorAlloc.txnSeqNumAllocator,
+			// The pipeliner is needed on leaves to ensure that in-flight writes
+			// are chained onto by reads that should see them.
+			&tcs.interceptorAlloc.txnPipeliner,
+			// The span refresher was configured above to not actually perform
+			// refreshes for leaves. It is still needed for accumulating the
+			// spans to be reported to the Root. But the gateway doesn't do much
+			// with them; see #24798.
+			&tcs.interceptorAlloc.txnSpanRefresher,
+		}
 		// All other interceptors are absent from a LeafTxn's interceptor stack
 		// because they do not serve a role on leaves.
 		tcs.interceptorStack = tcs.interceptorAlloc.arr[:3]

--- a/pkg/kv/txn_interceptor_committer.go
+++ b/pkg/kv/txn_interceptor_committer.go
@@ -52,8 +52,13 @@ func (tc *txnCommitter) SendLocked(
 		return tc.sendLockedWithElidedEndTransaction(ctx, ba, et)
 	}
 
-	// Assign the transaction's key to the Request's header.
-	et.Key = ba.Txn.Key
+	// Assign the transaction's key to the Request's header if it isn't already
+	// set. This is the only place where EndTransactionRequest.Key is assigned,
+	// but we could be dealing with a re-issued batch after a refresh. Remember,
+	// the committer is below the span refresh on the interceptor stack.
+	if et.Key == nil {
+		et.Key = ba.Txn.Key
+	}
 
 	// Pass the adjusted batch through the wrapped lockedSender.
 	return tc.wrapped.SendLocked(ctx, ba)

